### PR TITLE
DAOS-1452 utils: build and install with libunwind

### PR DESCRIPTION
**Pull request by bfaccini**
Now that libunwind usage has been released inside
Argobots to dump ULT stacks with symbols, build
Argobots with libunwind configured-in and makes
libunwind a Requires.

Also bump Argobots version/commit to the one that
has landed libunwind support in Argobots.

Change-Id: I3775c21005a8c42d865045ce58cdf97745a99415
Signed-off-by: Bruno Faccini <bruno.faccini@intel.com>

### CODE REVIEW

File: utils/rpms/daos.spec [hunk 1]
```diff
@@ -142,6 +142,8 @@ Requires(postun): /sbin/ldconfig
 Requires: libfabric >= 1.11.0
 %{?systemd_requires}
 Obsoletes: cart
+# libunwind for Argobots
+Requires: libunwind
```
Comment (JohnMalmberg) [inline utils/rpms/daos.spec:146]: The argobots package should be installing the libunwind package.
Unless daos directly requires libunwind without argobots, I don't think this dependency should be there.

The spec file and debian equivalent for argobots needs to be updated for the new version and the new dependencies.

**Review by JohnMalmberg — CHANGES_REQUESTED** [review]

File: utils/docker/Dockerfile.centos.7 [hunk 2]
```diff
@@ -78,6 +78,9 @@ RUN yum -y install \
            java-1.8.0-openjdk python-distro python36-distro
 RUN yum -y install openmpi3-devel
 
+# libunwind for Argobots
+RUN yum -y install libunwind libunwind-devel
```
Comment (brianjmurrell) [inline utils/docker/Dockerfile.centos.7:82]:
```suggestion
RUN yum -y install libunwind-devel
```

`libunwind-devel` will depend on `libunwind`.

File: utils/docker/Dockerfile.leap.15 [hunk 3]
```diff
@@ -99,6 +99,9 @@ RUN if [ -z "${REPO_GROUP_LEAP15}" ]; then
 # https://bugzilla.opensuse.org/show_bug.cgi?id=1178985
 RUN zypper --non-interactive install -y ipmctl-devel go1.14 go1.14-race
 
+# libunwind for Argobots
+RUN zypper --non-interactive install -y libunwind libunwind-devel
```
Comment (brianjmurrell) [inline utils/docker/Dockerfile.leap.15:103]:
```suggestion
RUN zypper --non-interactive install -y libunwind-devel
```

Ditto.

File: utils/docker/Dockerfile.ubuntu.20.04 [hunk 4]
```diff
@@ -37,6 +37,10 @@ RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y \
     apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y \
                        libsafec-dev libipmctl-dev ndctl ipmctl golang-go
 
+# libunwind for Argobots
+RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y \
+                       libunwind8 libunwind-dev
```
Comment (brianjmurrell) [inline utils/docker/Dockerfile.ubuntu.20.04:42]:
```suggestion
                       libunwind-dev
```
Ditto.

File: utils/sl/components/__init__.py [hunk 5]
```diff
@@ -287,7 +287,8 @@ def define_components(reqs):
                 commands=['git clean -dxf ',
                           './autogen.sh',
                           './configure --prefix=$ARGOBOTS_PREFIX CC=gcc'
-                          ' --enable-valgrind',
+                          ' --enable-valgrind'
+			  ' --with-libunwind=/usr/lib64 --enable-stack-unwind',
```
Comment (brianjmurrell) [inline utils/sl/components/__init__.py:291]: Does `--with-libunwind=` really require a path when it's a system standard path?  `configure` should usually be able to check standard system paths automatically. If it cannot, that sounds like a bug in argobots' `configure` and should be reported upstream.

**Review by brianjmurrell — CHANGES_REQUESTED** [review]

File: utils/rpms/daos.spec (hunk 1 above)
Comment (bfaccini) [reply utils/rpms/daos.spec:146]: Since original Argobots packaging is source only, I presume we create our own RPMs for it, so do you mean you will take the action ?? ;-)

File: utils/docker/Dockerfile.centos.7 (hunk 2 above)
Comment (bfaccini) [reply utils/docker/Dockerfile.centos.7:82]: Well, this is harmless, and thus why not be as explicit as possible since both are really needed respectively at run and build times ??

File: utils/sl/components/__init__.py (hunk 5 above)
Comment (bfaccini) [reply utils/sl/components/__init__.py:291]: Hmm, you are right Brian, I forgot to remove this change prior to push patch, BTW the path is even wrong!, sorry about this and thx to notice it !

File: utils/docker/Dockerfile.centos.7 (hunk 2 above)
Comment (brianjmurrell) [reply utils/docker/Dockerfile.centos.7:82]: It's not harmless.  The relationship between `libunwind-devel` and the other packages (i.e. `libunwind`) is at the sole discretion of the packager.  There is no guarantee that `libunwind` will always be the package that `libunwind-devel` depends on.  Depending on (only) `libunwind-devel` is the only assumption you can be assured of now and in the future, and the packager is then free to change the name, composition, or any other aspect of the `libunwind*` packages without impacting what consumers of `libunwind-devel` might be assuming.

TL; DR: Over-specifying dependencies can have all sorts of problems in the future that correctly specifying only the minimum dependency set prevents.

Indeed, we are addressing a bug in another package at this current time due to over-specifying dependencies.

File: utils/rpms/daos.spec (hunk 1 above)
Comment (brianjmurrell) [reply utils/rpms/daos.spec:146]: Yes, [`argobots.spec`](https://github.com/daos-stack/argobots/blob/master/argobots.spec) is where the dependency belongs.  That also needs updating to reflect the [update of version from 1.0 to 7fd19876b77f1f4f5787aed8cd242e60b4d653b6](https://github.com/daos-stack/daos/pull/4102/files#diff-d3984dc6b0c98ea42c422b465e317d22b0adc2c332dc6103254e3a72e0a1b37aR5)

The patch from v.10 to 7fd19876b77f1f4f5787aed8cd242e60b4d653b6 should be added to [`argobots.spec`](https://github.com/daos-stack/argobots/blob/master/argobots.spec) as well as adding the git hash to the [`argobots.spec` `Release:`](https://github.com/daos-stack/argobots/blob/b9a454393b7407e3b49d3db59881c1fcccff3c8e/argobots.spec#L3).

Ultimately every change to `utils/build.config` will require updating an RPM packaging of whatever was being updated in `utils/build.config`.

Comment (bfaccini) [reply utils/rpms/daos.spec:146]: Ok, but correct me if I am wrong, this means to also push a separate PR to this "Argobots packaging" GitHub project/repo, right ??
And I believe, I must also add the "configure --enable-libunwind" step into Argobots.spec too.

File: utils/docker/Dockerfile.centos.7 (hunk 2 above)
Comment (bfaccini) [reply utils/docker/Dockerfile.centos.7:82]: Ok...

File: utils/rpms/daos.spec (hunk 1 above)
Comment (brianjmurrell) [reply utils/rpms/daos.spec:146]: > Ok, but correct me if I am wrong, this means to also push a separate PR to this "Argobots packaging" GitHub project/repo, right ??

That is correct.

> And I believe, I must also add the "configure --enable-libunwind" step into Argobots.spec too.

Indeed, that is also correct.  Some notes about the [daos-stack/argobots](https://github.com/daos-stack/argobots) updates:
- that package is currently on version 1.0 of argobots: https://github.com/daos-stack/argobots/blob/b9a454393b7407e3b49d3db59881c1fcccff3c8e/argobots.spec#L2
  - but upstream as released 1.0.1 and indeed 1.1a1
  - presumably the git commit you want is actually after 1.1a1
  - if that is so, then you should update the argobots package to 1.1a1 and then calculate the diff between 1.1a1 and the commit you want to use and add that as a `Patch1:` to the specfile
  - you will also need to add `.g7fd1987` to the `Release:  1%{?dist}`: https://github.com/daos-stack/argobots/blob/b9a454393b7407e3b49d3db59881c1fcccff3c8e/argobots.spec#L3
 
You will have to make similar changes to `debian/` in that repo.

I can help you as you get into the details.

**Review by brianjmurrell — DISMISSED** [review]

**Comment by brianjmurrell** [comment]
Adding @jolivier23 and @johannlombardi just to make sure this version of argobots is acceptable.

You also need to merge master.  You have a couple of conflicts.

File: utils/sl/components/__init__.py [hunk 6]
```diff
@@ -287,7 +287,8 @@ def define_components(reqs):
                 commands=['git clean -dxf ',
                           './autogen.sh',
                           './configure --prefix=$ARGOBOTS_PREFIX CC=gcc'
-                          ' --enable-valgrind',
+                          ' --enable-valgrind'
+			  ' --enable-stack-unwind',
```
Comment (jolivier23) [inline utils/sl/components/__init__.py:291]: Are you using tabs here?  When I select this line, it looks like tabs...It should be spaces in python file

Comment (bfaccini) [reply utils/sl/components/__init__.py:291]: Oops, good catch and thanks !

**Review by jolivier23 — DISMISSED** [review]
Probably would be good for someone to check this version for any regressions...

**Comment by bfaccini** [comment]
Hmm, looks like the last merge with recent master has brought up a new docker file for CentOS 8 ....

**Comment by bfaccini** [comment]
> Probably would be good for someone to check this version for any regressions...

Could we consider that the 32 successful checks are already a good start to verify against any regression ? May be some performance check is also needed ...

**Comment by brianjmurrell** [comment]
> Probably would be good for someone to check this version for any regressions...

Yes, completely agreed.  My main driver for adding you as a reviewer was an approval of updating to the new release.  If you know of anyone else that might have a strong opinion please do add them as a reviewer here.

**Review by brianjmurrell — COMMENTED** [review]
LGTM, but I'd rather leave my +1 slot for somebody with more opinion on moving to this new release.

**Review by jolivier23 — DISMISSED** [review]

**Comment by brianjmurrell** [comment]
@bfaccini Looks like you have conflicts that need to be resolved here.  Can you merge master and resolve those and push a new commit?

File: utils/build.config [hunk 7]
```diff
@@ -2,7 +2,7 @@
 component=daos
 
 [commit_versions]
-ARGOBOTS = v1.0
+ARGOBOTS = 7fd19876b77f1f4f5787aed8cd242e60b4d653b6
```
Comment (ashleypittman) [inline utils/build.config:5]: As this is moving away from a release version of argobots should this be a post 1.2 feature?   Otherwise we need to audit what else we're picking up here.

**Review by ashleypittman — CHANGES_REQUESTED** [review]
This would be great to have but we need to address the comments and decide if this is for 1.2 or post 1.2

File: Jenkinsfile [hunk 8]
```diff
@@ -154,7 +154,8 @@ String unit_packages() {
                            'libipmctl-devel python36-pyxattr ' +
                            'python36-tabulate numactl ' +
                            'libyaml-devel ' +
-                           'valgrind-devel patchelf'
+                           'valgrind-devel patchelf ' +
+                           'libunwind'
```
[the remark this answers is not in the document]
Comment (brianjmurrell) [reply Jenkinsfile:158]: > If this line really is needed then we need to report that to the argobots maintainers.

The argobots packagers, to be precise, which is us.  :-)  But you are right @ashleypittman this is done automatically by the inclusion of the `argobots` package on L147 and so @bruno the change here is unnecessary:
```suggestion
```

File: utils/build.config (hunk 7 above)
Comment (brianjmurrell) [reply utils/build.config:5]: The alternative (and in fact the generally preferred approach in the packaging world) is to extract just the libunwind patch from argobots' master and apply it in isolation of everything else that has landed to argobots' master since 1.1.

File: Jenkinsfile (hunk 8 above)
Comment (brianjmurrell) [inline Jenkinsfile:158]:
```suggestion
                           'valgrind-devel patchelf'
```

**Review by brianjmurrell — CHANGES_REQUESTED** [review]

File: Jenkinsfile (hunk 8 above)
Comment (bfaccini) [reply Jenkinsfile:158]: @brianjmurrell and @ashleypittman , do you mean that in CI there is no more local build of DAOS but only RPM-based installs ?? 
Then I may have to add libunwind detection in utils/sl/components/__init__.py and only configure/build Argobots with libunwind support if present and print a warning message if not ?

Comment (brianjmurrell) [reply Jenkinsfile:158]: No.  To clarify, `unit_packages()` is a list of packages to install during unit tests to which you added `libunwind`.  However that list already includes `argobots`: https://github.com/daos-stack/daos/blob/792a88a4c31784146c7e4740f070d7d653d9f874/Jenkinsfile#L147

That `argobots` reference will automatically include `libunwind` without it needing to be named explicitly here.

Comment (ashleypittman) [reply Jenkinsfile:158]: CI will do either local build or RPM based builds depending on if you set Quick-build or not.

Default is local build, and for that you need to ensure the package is in the dockerfile which you've done, if you set Quick-build:true then it'll use rpms and for that the argobots rpm needs to be built correctly, either way it should use the same features so no need to modify components at all.

**Review by ashleypittman — COMMENTED** [review]
Code looks fine, just a case of deciding which commit/release we want to ship with.

**Comment by ashleypittman** [comment]
NLT failed with 01:04:54.574  ERROR: daos_engine:0 /localhome/jenkins/build/install/bin/daos_engine: error while loading shared libraries: libunwind.so.8: cannot open shared object file: No such file or directory, so I guess that's our argobots rpm packaging?

I don't know how the software is installed for NLT and unit-test VMs.

**Comment by bfaccini** [comment]
> so I guess that's our argobots rpm packaging?

I guess too
> I don't know how the software is installed for NLT and unit-test VMs.

Should be RPMs

So I have added the "PR-repos: argobots@PR-16" pragma in order for this PR to use the one with associated argots-packaging changes.

**Comment by bfaccini** [comment]
> > so I guess that's our argobots rpm packaging?
> 
> I guess too
> 
> > I don't know how the software is installed for NLT and unit-test VMs.
> 
> Should be RPMs
> 
> So I have added the "PR-repos: argobots@PR-16" pragma in order for this PR to use the one with associated argobots-packaging changes.

Well the error remains even if argobots@PR-16 packaging PR, to build Argobots RPMs with stack unwinding support and libunwind[-devel] require, is enforced.
Since this failures have started to happen when "libunwind" pkg forced/explicit installation has been removed from Jenkinsfiles (assuming that it will occur as a pre-req of Argobots RPM), @brianjmurrell could it be because the argobots RPM being used is the one available on our repos which is still an older version ?

**Comment by brianjmurrell** [comment]
`argbots` is being installed from the PR:

```
[2021-02-14T13:35:05.439Z]  argobots            x86_641.1~a1-1.g7fd1987.el7 build.hpdd.intel.com_job_daos-stack_job_argobots_job_PR-16_lastSuccessfulBuild_artifact_artifacts_centos7_ 62 k
```

But that install group does not include `libunwind`:

```
[2021-02-14T13:35:05.438Z] ================================================================================
[2021-02-14T13:35:05.438Z]  Package             ArchVersion               Repository                                                                                                 Size
[2021-02-14T13:35:05.438Z] ================================================================================
[2021-02-14T13:35:05.438Z] Installing:
[2021-02-14T13:35:05.438Z]  libfabric-devel     x86_641.11.1-1.el7          repo.dc.hpdd.intel.com_repository_daos-stack-el-7-x86_64-stable-local_                                     209 k
[2021-02-14T13:35:05.438Z]  libisa-l-devel      x86_642.30.0-1.el7          repo.dc.hpdd.intel.com_repository_daos-stack-el-7-x86_64-stable-local_                                     125 k
[2021-02-14T13:35:05.438Z]  libpmem             x86_641.9.2-1.el7           repo.dc.hpdd.intel.com_repository_daos-stack-el-7-x86_64-stable-local_                                     81 k
[2021-02-14T13:35:05.438Z]  libpmemobj          x86_641.9.2-1.el7           repo.dc.hpdd.intel.com_repository_daos-stack-el-7-x86_64-stable-local_                                     141 k
[2021-02-14T13:35:05.438Z]  protobuf-c          x86_641.3.3-1.el7           repo.dc.hpdd.intel.com_repository_daos-stack-el-7-x86_64-stable-local_                                     31 k
[2021-02-14T13:35:05.438Z]  spdk-devel          x86_6420.01.1-1.el7         repo.dc.hpdd.intel.com_repository_daos-stack-el-7-x86_64-stable-local_                                     845 k
[2021-02-14T13:35:05.438Z]  gotestsum           x86_640.5.4-1.el7           repo.dc.hpdd.intel.com_repository_daos-stack-el-7-x86_64-stable-local_                                     1.6 M
[2021-02-14T13:35:05.438Z]  pmix                x86_642.1.1-3.el7           repo.dc.hpdd.intel.com_repository_daos-stack-el-7-x86_64-stable-local_                                     3.0 M
[2021-02-14T13:35:05.438Z]  boost-devel         x86_641.53.0-28.el7         repo.dc.hpdd.intel.com_repository_centos-7.9-x86_64-group_                                                 7.0 M
[2021-02-14T13:35:05.438Z]  hwloc-devel         x86_641.11.8-4.el7          repo.dc.hpdd.intel.com_repository_centos-7.9-x86_64-group_                                                 208 k
[2021-02-14T13:35:05.438Z]  libyaml-devel       x86_640.1.4-11.el7_0        repo.dc.hpdd.intel.com_repository_centos-7.9-x86_64-group_                                                 82 k
[2021-02-14T13:35:05.438Z]  numactl             x86_642.0.12-5.el7          repo.dc.hpdd.intel.com_repository_centos-7.9-x86_64-group_                                                 66 k
[2021-02-14T13:35:05.438Z]  numactl-devel       x86_642.0.12-5.el7          repo.dc.hpdd.intel.com_repository_centos-7.9-x86_64-group_                                                 24 k
[2021-02-14T13:35:05.438Z]  openmpi3            x86_643.1.3-2.el7           repo.dc.hpdd.intel.com_repository_centos-7.9-x86_64-group_                                                 2.9 M
[2021-02-14T13:35:05.438Z]  valgrind-devel      x86_641:3.15.0-11.el7       repo.dc.hpdd.intel.com_repository_centos-7.9-x86_64-group_                                                 2.4 M
[2021-02-14T13:35:05.438Z]  fuse3               x86_643.6.1-4.el7           repo.dc.hpdd.intel.com_repository_centos-7.9-x86_64-group_                                                 47 k
[2021-02-14T13:35:05.438Z]  fuse3-libs          x86_643.6.1-4.el7           repo.dc.hpdd.intel.com_repository_centos-7.9-x86_64-group_                                                 82 k
[2021-02-14T13:35:05.438Z]  libipmctl-devel     x86_6402.00.00.3833-1.el7   repo.dc.hpdd.intel.com_repository_centos-7.9-x86_64-group_                                                 34 k
[2021-02-14T13:35:05.438Z]  patchelf            x86_640.12-1.el7            repo.dc.hpdd.intel.com_repository_centos-7.9-x86_64-group_                                                 86 k
[2021-02-14T13:35:05.438Z]  python36-pyxattr    x86_640.6.1-1.el7           repo.dc.hpdd.intel.com_repository_centos-7.9-x86_64-group_                                                 27 k
[2021-02-14T13:35:05.438Z]  python36-tabulate   noarch0.8.3-8.el7           repo.dc.hpdd.intel.com_repository_centos-7.9-x86_64-group_                                                 44 k
[2021-02-14T13:35:05.439Z]  argobots            x86_641.1~a1-1.g7fd1987.el7 build.hpdd.intel.com_job_daos-stack_job_argobots_job_PR-16_lastSuccessfulBuild_artifact_artifacts_centos7_ 62 k
[2021-02-14T13:35:05.439Z] Upgrading:
[2021-02-14T13:35:05.439Z]  libfabric           x86_641.11.1-1.el7          repo.dc.hpdd.intel.com_repository_daos-stack-el-7-x86_64-stable-local_                                     553 k
[2021-02-14T13:35:05.439Z]  ipmctl              x86_6402.00.00.3833-1.el7   repo.dc.hpdd.intel.com_repository_centos-7.9-x86_64-group_                                                 81 k
[2021-02-14T13:35:05.439Z]  libipmctl           x86_6402.00.00.3833-1.el7   repo.dc.hpdd.intel.com_repository_centos-7.9-x86_64-group_                                                 477 k
[2021-02-14T13:35:05.439Z] Installing dependencies:
[2021-02-14T13:35:05.439Z]  dpdk                x86_6419.11-1.el7           repo.dc.hpdd.intel.com_repository_daos-stack-el-7-x86_64-stable-local_                                     256 k
[2021-02-14T13:35:05.439Z]  dpdk-devel          x86_6419.11-1.el7           repo.dc.hpdd.intel.com_repository_daos-stack-el-7-x86_64-stable-local_                                     252 k
[2021-02-14T13:35:05.439Z]  libisa-l            x86_642.30.0-1.el7          repo.dc.hpdd.intel.com_repository_daos-stack-el-7-x86_64-stable-local_                                     101 k
[2021-02-14T13:35:05.439Z]  spdk                x86_6420.01.1-1.el7         repo.dc.hpdd.intel.com_repository_daos-stack-el-7-x86_64-stable-local_                                     764 k
[2021-02-14T13:35:05.439Z]  libpsm2-devel       x86_6411.2.78-1.el7         repo.dc.hpdd.intel.com_repository_centos-7.9-x86_64-group_                                                 56 k
[2021-02-14T13:35:05.439Z]  libuuid-devel       x86_642.23.2-65.el7         repo.dc.hpdd.intel.com_repository_centos-7.9-x86_64-group_                                                 93 k
[2021-02-14T13:35:05.439Z] 
[2021-02-14T13:35:05.439Z] Transaction Summary
[2021-02-14T13:35:05.439Z] ================================================================================
[2021-02-14T13:35:05.439Z] Install  28 Packages
[2021-02-14T13:35:05.439Z] Upgrade   3 Packages
```

So there does indeed seem to be a problem with the argobots PR.

The problem seems to be the use and _belief_ that:

```
%global configopts
```

Is used in `%configure` automatically.  It's not.  This would seem to be probably due to using the libfabric specfile as a template for the argobots one and not fully removing the `configopts` support from it.  In order to use `configopts` it also needs to be added to the `%configure` line.  I'll suggest an edit or two to resolve this...

**Review by jolivier23 — APPROVED** [review]

**Comment by brianjmurrell** [comment]
@bfaccini Looks like your last commit here forgot to include `PR-repos: argobots@PR-16`.

You can use `git commit --allow-empty` to create a new empty commit with the above in it if you wish.

**Comment by bfaccini** [comment]
> @bfaccini Looks like your last commit here forgot to include `PR-repos: argobots@PR-16`.
> 
> You can use `git commit --allow-empty` to create a new empty commit with the above in it if you wish.

Oops, you are right, I forgot that commit tags need to be explicitly added for each new commit push ... Thanks again Brian!

**Review by jolivier23 — APPROVED** [review]

**Review by ashleypittman — APPROVED** [review]

**Review by brianjmurrell — APPROVED** [review]
